editoast/src/core/mq_client.rs

```diff
@@ -0,0 +1,232 @@
+use editoast_derive::EditoastError;
+use futures_util::StreamExt;
+use itertools::Itertools;
+use lapin::{
+    options::{BasicConsumeOptions, BasicPublishOptions},
+    types::{ByteArray, FieldTable, ShortString},
+    BasicProperties, Connection, ConnectionProperties,
+};
+use serde::Serialize;
+use serde_json::to_vec;
+use std::{fmt::Debug, sync::Arc};
+use thiserror::Error;
+use tokio::time::{timeout, Duration};
+
+#[derive(Debug, Clone)]
+pub struct RabbitMQClient {
+    connection: Arc<Connection>,
+    exchange: String,
+    timeout: u64,
+    hostname: String,
+}
+
+pub struct Options {
+    /// format `amqp://username:password@host:port/vhost`
+    /// for instance: `amqp://osrd:password@localhost:5672/%2f` for the default vhost
+    pub uri: String,
+    /// Exchange name
+    pub worker_pool_identifier: String,
+    /// Default timeout for the response
+    pub timeout: u64,
+}
+
+#[derive(Debug, Error, EditoastError)]
+#[editoast_error(base_id = "coreclient")]
+pub enum Error {
+    #[error("AMQP error: {0}")]
+    #[editoast_error(status = "500")]
+    Lapin(#[from] lapin::Error),
+    #[error("Cannot serialize request: {0}")]
+    #[editoast_error(status = "500")]
+    Serialization(serde_json::Error),
+    #[error("Cannot parse response status")]
+    #[editoast_error(status = "500")]
+    StatusParsing,
+    #[error("Response timeout")]
+    #[editoast_error(status = "500")]
+    ResponseTimeout,
+}
+
+pub struct MQResponse {
+    pub payload: Vec<u8>,
+    pub status: Vec<u8>,
+}
+
+impl RabbitMQClient {
+    pub async fn new(options: Options) -> Result<Self, Error> {
+        let connection = Connection::connect(&options.uri, ConnectionProperties::default())
```

Comment:
Naive question, don't we want a pool of connection instead of a unique connection?

Reply:
The preferred mechanism is using channels inside an unique connection (see https://www.rabbitmq.com/tutorials/amqp-concepts#amqp-channels). This gives use multiplexing at a protocol level.

However, it uses a single TCP connection, so it suffers from the usual problems (head of line blocking, etc). It don't think those will come up at our planned deployment scale.